src/status.md

```diff
@@ -0,0 +1,313 @@
+
+### Generics
+
+AssemblyScript compiles generics to one concrete method or function per set of unique contextual type arguments, known as monomorphization. Implications are that a module only includes and exports concrete functions for sets of type arguments actually used and that concrete functions can be shortcutted with [static type checks](./environment.md#static-type-checks) at compile time, which turned out to be quite useful.
+
+* The compiler does not currently enforce `extends Y` clauses on type parameters. Likely to be enforced in the future.
+* WebAssembly GC 🦄 may introduce more sophisticated mechanisms like reified generics, potentially post-MVP.
+* Concrete functions compiling to the exact same code are de-duplicated during optimization.
+
+### Garbage collection
+
+[Garbage collection](./garbage-collection.md) is currently implemented in linear memory, independent from the host, and is best paired with [the loader](./loader.md) for interop.
+
+### Interop with JS
+
+WebAssembly only understands numeric values as of today and cannot easily exchange objects with JavaScript. Hence, when an object is returned from WebAssembly to JavaScript, what the caller passes and the callee receives is a pointer to the object in linear memory. Note that WebAssembly does not know what to do with JavaScript objects passed to it, as the VM will implicitly convert the object to a number when it crosses the boundary, which is typically not what you want.
+
+For example, to pass a string to a WebAssembly export, one first has to allocate the string in the WebAssembly module's linear memory, and then pass the resulting pointer to the WebAssembly export. The same is true for arrays and other objects.
+
+For now, [the loader](./loader.md) provides the utility necessary to translate between objects in linear memory and JavaScript objects (e.g. with `__newString` and `__getString`), and our hopes are on Type Imports 🦄, Interface Types 🦄 and perhaps Garbage collection 🦄 to make interop more convenient eventually.
+
+See also: [Will interop between AssemblyScript and JavaScript become better?](./frequently-asked-questions.md#will-interop-between-assemblyscript-and-javascript-become-better)
+
+### Union types
+
+WebAssembly cannot efficiently represent locals or globals of dynamic types, so union types are not supported in AssemblyScript. One can however take advantage of the fact that AssemblyScript is a static compiler, with monomorphized generics and [static type checks](./environment.md#static-type-checks), to achieve a similar effect:
+
+```ts
+function addOrConcat<T>(a: T, b: T): T {
+  return a + b; // concats if a string, otherwise adds
+}
+
+function addOrSomethingElse<T>(a: T, b: T): T {
+  if (isString<T>()) {
+    return "something else"; // eliminated if T is not a string
+  } else {
```

Comment:
This often works, but only if it can be proven statically, for example on self- or downcasts. Otherwise one gets a dynamic check there, with an actual branch.